miservice/minaservice.py

```diff
@@ -105,11 +116,47 @@ async def player_set_loop(self, deviceId, type=1):
         )
 
     async def play_by_url(self, deviceId, url, _type=2):
+        if deviceId not in self.device2hardware:
+            await self._init_devices()
+        hardware = self.device2hardware[deviceId]
+        if hardware in _USE_PLAY_MUSIC_API:
+            return await self.play_by_music_url(deviceId, url, _type)
+        else:
+            return await self.ubus_request(
+                deviceId,
+                "player_play_url",
+                "mediaplayer",
+                {"url": url, "type": _type, "media": "app_ios"},
+            )
+
+    async def _init_devices(self):
+        hardware_data = await self.device_list()
+        for h in hardware_data:
+            deviceId = h.get("deviceID", "")
+            hardware = h.get("hardware", "")
+            self.device2hardware[deviceId] = hardware
```

Comment:
if deviceId and hardware:
    self.device2hardware[deviceId] = hardware
    break